compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp

```diff
@@ -958,6 +958,88 @@ class MoveAllDependenciesIntoOp : public OpRewritePattern<OpTy> {
   }
 };
 
+/// Utility to convert function arguments with memref element types' bitwidth
+/// either 8 or 16 to i32. This basically involves linearizing an N-d memref to
+/// 1-d i32 memref and adjusting the offset/size/stride metadata.
+static LogicalResult CastFunctionArgs(func::FuncOp funcOp,
+                                      PatternRewriter &rewriter) {
+  // only run on npu control functions
+  bool hasNpuOps = false;
+  funcOp.walk([&](AIEX::NpuDmaMemcpyNdOp dma) { hasNpuOps = true; });
+  if (!hasNpuOps) return failure();
+
+  // cast all the function args to i32 types.
+  // this is in support of npu.dma_memcpy_nd which only allow 32bit types
+  mlir::FunctionType funcType = funcOp.getFunctionType();
+  SmallVector<Type> argTypes(funcType.getInputs());
+  for (int i = 0, e = argTypes.size(); i < e; i++) {
+    auto memrefTy = dyn_cast<MemRefType>(argTypes[i]);
+    if (!memrefTy) continue;
+
+    unsigned int bitwidth = memrefTy.getElementTypeBitWidth();
+    if (bitwidth != 16 && bitwidth != 8) continue;
+
+    unsigned int div = 32 / bitwidth;
+    unsigned int numElements = memrefTy.getNumElements() / div;
+    SmallVector<int64_t> shape{numElements};
+    MemRefType newMemrefTy =
+        MemRefType::get(shape, rewriter.getIntegerType(32));
+    argTypes[i] = newMemrefTy;
+    auto &entry = funcOp.front();
+    entry.insertArgument(i, newMemrefTy, rewriter.getUnknownLoc());
+    rewriter.setInsertionPointToStart(&entry);
+    // With memref shape collapsed to 1d, the multi-dimensional offset also
+    // needs to be collapsed.
```

Comment:
The memref rank changing  should not determine what we do to the offsets (same with sizes/strides)  is independent of the rank. I believe we need to divide any stride 1 offsets by `div` and can leave everything as is?

Reply:
I tried applying that logic but it didn't work.

This is the first row :-
```
64x64xf32=[384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256]
```

Reply:
to me it looks like you are doing nothing to the offsets in the linearization, so I wonder if leaving them unmodified, would work?